File: android/app/src/main/java/com/videoconverter/android/domain/ProbeParser.kt
package com.videoconverter.android.domain

import org.json.JSONObject

fun parseFfprobeJson(sourceUri: String, displayName: String, json: String): MediaInfo =
    try {
        mapProbe(sourceUri, displayName, JSONObject(json))
    } catch (_: Exception) {
        unreadable(sourceUri, displayName, "无法解析媒体信息")
    }

fun unreadable(sourceUri: String, displayName: String, reason: String): MediaInfo =
    MediaInfo(
        sourceUri = sourceUri,
        displayName = displayName,
        importable = false,
        error = reason,
    )

private fun mapProbe(sourceUri: String, displayName: String, probe: JSONObject): MediaInfo {
    val format = probe.optJSONObject("format")
    val streams = probe.optJSONArray("streams")
    var video: JSONObject? = null
    var audio: JSONObject? = null

    if (streams != null) {
        for (index in 0 until streams.length()) {
            val stream = streams.optJSONObject(index) ?: continue
            when (stream.optionalString("codec_type")) {
                "video" -> if (video == null) video = stream
                "audio" -> if (audio == null) audio = stream
            }
        }
    }

    val hasAv = video != null || audio != null
    return MediaInfo(
        sourceUri = sourceUri,
        displayName = displayName,
        durationSecs = format?.optionalString("duration")?.toDoubleOrNull(),
        container = format?.optionalString("format_name"),
        videoCodec = video?.optionalString("codec_name"),
        width = video?.optionalInt("width"),
        height = video?.optionalInt("height"),
        frameRate = parseFrameRate(video?.optionalString("r_frame_rate")),
        audioCodec = audio?.optionalString("codec_name"),
        channels = audio?.optionalInt("channels"),
        importable = hasAv,
        error = if (hasAv) null else "没有可转码的视频或音频流",
    )
}

private fun parseFrameRate(value: String?): Double? {
    value ?: return null
    val separator = value.indexOf('/')
    if (separator < 0) return value.toDoubleOrNull()

    val numerator = value.substring(0, separator).toDoubleOrNull() ?: return null
    val denominator = value.substring(separator + 1).toDoubleOrNull() ?: return null
    return if (denominator == 0.0) null else numerator / denominator
}

private fun JSONObject.optionalString(name: String): String? =
    if (has(name) && !isNull(name)) optString(name) else null

private fun JSONObject.optionalInt(name: String): Int? =
    if (has(name) && !isNull(name)) optInt(name) else null

File: ios/LiteTrans/Domain/ProbeParser.swift
import Foundation

public func parseFfprobeJson(
    sourceUri: String,
    displayName: String,
    json: String,
    cannotParse: String = "Could not parse media information",
    noAvStream: String = "No convertible video or audio stream"
) -> MediaInfo {
    do {
        guard let data = json.data(using: .utf8) else {
            return unreadable(sourceUri: sourceUri, displayName: displayName, reason: cannotParse)
        }
        let object = try JSONSerialization.jsonObject(with: data)
        guard let probe = object as? [String: Any] else {
            return unreadable(sourceUri: sourceUri, displayName: displayName, reason: cannotParse)
        }
        return mapProbe(sourceUri: sourceUri, displayName: displayName, probe: probe, noAvStream: noAvStream)
    } catch {
        return unreadable(sourceUri: sourceUri, displayName: displayName, reason: cannotParse)
    }
}

public func unreadable(sourceUri: String, displayName: String, reason: String) -> MediaInfo {
    MediaInfo(
        sourceUri: sourceUri,
        displayName: displayName,
        importable: false,
        error: reason
    )
}

private func mapProbe(
    sourceUri: String,
    displayName: String,
    probe: [String: Any],
    noAvStream: String
) -> MediaInfo {
    let format = probe["format"] as? [String: Any]
    let streams = probe["streams"] as? [[String: Any]]
    var video: [String: Any]?
    var audio: [String: Any]?

    if let streams {
        for stream in streams {
            switch optionalString(stream, "codec_type") {
            case "video":
                if video == nil { video = stream }
            case "audio":
                if audio == nil { audio = stream }
            default:
                break
            }
        }
    }

    let hasAv = video != nil || audio != nil
    return MediaInfo(
        sourceUri: sourceUri,
        displayName: displayName,
        durationSecs: format.flatMap { optionalString($0, "duration") }.flatMap(Double.init),
        container: format.flatMap { optionalString($0, "format_name") },
        videoCodec: video.flatMap { optionalString($0, "codec_name") },
        width: video.flatMap { optionalInt($0, "width") },
        height: video.flatMap { optionalInt($0, "height") },
        frameRate: parseFrameRate(video.flatMap { optionalString($0, "r_frame_rate") }),
        audioCodec: audio.flatMap { optionalString($0, "codec_name") },
        channels: audio.flatMap { optionalInt($0, "channels") },
        importable: hasAv,
        error: hasAv ? nil : noAvStream
    )
}

private func parseFrameRate(_ value: String?) -> Double? {
    guard let value else { return nil }
    guard let separator = value.firstIndex(of: "/") else {
        return Double(value)
    }

    guard let numerator = Double(value[..<separator]),
          let denominator = Double(value[value.index(after: separator)...])
    else {
        return nil
    }
    return denominator == 0.0 ? nil : numerator / denominator
}

private func optionalString(_ object: [String: Any], _ name: String) -> String? {
    guard let value = object[name], !(value is NSNull) else { return nil }
    if let string = value as? String { return string }
    if let number = value as? NSNumber { return number.stringValue }
    return nil
}

private func optionalInt(_ object: [String: Any], _ name: String) -> Int? {
    guard let value = object[name], !(value is NSNull) else { return nil }
    if let int = value as? Int { return int }
    if let number = value as? NSNumber { return number.intValue }
    if let string = value as? String { return Int(string) }
    return nil
}
